Channel for Google hangouts chat integration

muelerma opened the pull request:
**Proposed changes**:
- Please see issue #5006 for details

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=5060) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=5060) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=5060) it.</sub>

erohmensing commented:
Thanks for the PR! We'll give it a review as soon as possible.

degiz commented on tests/core/test_channels.py:
```diff
@@ -911,3 +911,210 @@ def test_has_user_permission_to_send_messages_to_conversation_without_permission
     )
+
+
+def test_hangouts_channel():
```
Thanks a lot for adding tests!

One optional thing: could you please move hangouts related tests to tests/core/channels/hangouts.py?
It's one big test file now and it becomes quite difficult to follow.

muelerma replied:
Done that

degiz commented on rasa/core/channels/hangouts.py:
```diff
@@ -0,0 +1,288 @@
+            except KeyError:
+                logger.error(
+                    "Buttons must be a list of dicts with 'title' and 'payload' as keys"
+                )
```
does it make sense to whether `pass` or `return` after Exception?

muelerma replied:
Added return

degiz commented on rasa/core/channels/hangouts.py:
```diff
@@ -0,0 +1,288 @@
+
+    @classmethod
+    def name(cls) -> Text:
+        return "hangouts"
```
this name appears twice, maybe we can extract it to constant?

muelerma replied:
Here I orientated myself on the other channels like facebook.py, where both InputChannel and OuputChanel need a name() classmethod and these also return the same value.

degiz replied:
you are right! I just meant that `hangouts` itself could be extracted to a const, and name() will return this constant.

muelerma replied:
Done that

degiz replied:
that's exactly it, thanks!

muelerma commented:
Hello Degiz,
thanks for your comments. I implemented most of the changes, except the credentials one, please see the comment. Additionally, Google is deprecating token based authorization for OAuth2, which I now added to the channel.

muelerma commented on rasa/core/channels/hangouts.py:
```diff
@@ -0,0 +1,288 @@
+import asyncio
```
Cleaned the imports (were initially just copied from channel.py)

muelerma commented on docs/user-guide/connectors/hangouts.rst:
```diff
@@ -0,0 +1,68 @@
+Running On Hangouts Chat
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you want to connect to Facebook using the run script, e.g. using:
```
Yes, fixed that.

muelerma commented on rasa/core/channels/hangouts.py:
```diff
@@ -0,0 +1,288 @@
+            credentials.get("user_added_intent"),
+            credentials.get("room_added_intent"),
+            credentials.get("removed_intent"),
+        )
```
Just to make sure I understand this right, I should add the from_configuration() method to InputChannel (why not only HangoutsChannel?) and this method would read a file called "hangouts_config.yml" with the intent specified in it?

degiz replied:
Sorry for confusion,

I meant that inside HangoutsChannel you can have fields like `hangouts_user_added_intent _name`, and so on.

degiz replied:
And this fields will contain actual Strings that you want to use.

muelerma replied:
Hi Degiz, thanks, I think I got it now. You mean the values should rather be hard-coded in the hangouts.py. When you say fields, do you mean attributes of the HangoutsInput class?

muelerma replied:
Please see the latest commit. Thanks your support!

degiz replied:
thanks, I meant exactly that!

tmbo commented:
@muelerma we'd love to get this in the next release (next tuesday). do you have time to look at the suggestions from @degiz to make sure we are ready to go?

muelerma commented:
@tmbo ok, nice to hear, I will work on it over the weekend.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=5060) <br/>All committers have signed the CLA.

muelerma commented:
> Hey @muelerma
> 
> thanks for answering the comments! I've just requested two more changes, and we'll be ready to merge!
> 
> about the OAuth2: I've tested the channel without it, and it seems to work fine. Google was using the webhook without any auth. Maybe we can just then remove mentions of OAuth from the description?

Hi @degiz 
which description do you mean, the `/docs/user-guide/connectors/hangouts.rst`? Modified that, please have a look. 

Yes, Google just sends the token, it is up to the bot implementation to check it, therefore it also works without it. If we want the token verification to be optional, I think it is good practice to make the `project_id` parameter, which is required for OAuth to work, also optional. That's why I modified `from_credentials` and constructor method of `HangoutsInput`, please let me know what you think.
